core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java

```diff
@@ -194,4 +194,196 @@ public void testNoBuffering() throws Exception {
+				"                 sh:resultPath                 ex:knows ;\n" +
+				"                 sh:resultSeverity             sh:Violation ;\n" +
+				"                 sh:sourceConstraintComponent  sh:NotConstraintComponent ;\n" +
+				"                 sh:sourceShape                _:b0 ;\n" +
+				"                 sh:value                      ex:steve\n" +
+				"               ] ;\n" +
+				"  sh:result    [ a                             sh:ValidationResult ;\n" +
+				"                 sh:focusNode                  ex:validPerson1 ;\n" +
+				"                 sh:resultPath                 ex:knows ;\n" +
+				"                 sh:resultSeverity             sh:Violation ;\n" +
+				"                 sh:sourceConstraintComponent  sh:NotConstraintComponent ;\n" +
+				"                 sh:sourceShape                _:b0 ;\n" +
+				"                 sh:value                      ex:steve\n" +
+				"               ]\n" +
+				"] .\n";
+
+		System.out.println("### EXPECTED ###");
+		System.out.println(data);
+		System.out.println("#################\n");
+
+		Model expected = Rio.parse(new StringReader(data), "", RDFFormat.TURTLE);
+
+		StringWriter stringWriter = new StringWriter();
+		WriterConfig config = new WriterConfig();
+		config.set(BasicWriterSettings.INLINE_BLANK_NODES, true);
+		Rio.write(expected, stringWriter, RDFFormat.TURTLE, config);
+
+		System.out.println("### ACTUAL ###");
+		System.out.println(stringWriter.toString());
+		System.out.println("#################\n");
+
+		Model actual = Rio.parse(new StringReader(stringWriter.toString()), "", RDFFormat.TURTLE);
+		assertTrue(Models.isomorphic(expected, actual));
+	}
+
+	@Test
+	public void anotherBnodeTestReduced() throws Exception {
+		String data = "" +
+				"@prefix ex:    <http://example.com/ns#> .\n" +
+				"\n" +
+				"ex:a  ex:list   (_:b0 _:b0) .";
+
+		System.out.println("### EXPECTED ###");
+		System.out.println(data);
+		System.out.println("#################\n");
+
+		Model expected = Rio.parse(new StringReader(data), "", RDFFormat.TURTLE);
+
+		StringWriter stringWriter = new StringWriter();
+		WriterConfig config = new WriterConfig();
+		config.set(BasicWriterSettings.INLINE_BLANK_NODES, true);
+		Rio.write(expected, stringWriter, RDFFormat.TURTLE, config);
+
+		System.out.println("### ACTUAL ###");
+		System.out.println(stringWriter.toString());
+		System.out.println("#################\n");
+
+		Model actual = Rio.parse(new StringReader(stringWriter.toString()), "", RDFFormat.TURTLE);
+		assertTrue(Models.isomorphic(expected, actual));
+	}
+
+	@Test
+	public void anotherBnodeTestReduced2() throws Exception {
+		String data = "@prefix ex:    <http://example.com/ns#> .\n" +
+				"@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .\n" +
+				"\n" +
+				"ex:ex1    ex:list  [ rdf:first  [  ] ;\n" +
+				"                             rdf:rest   rdf:nil ;\n" +
+				"                             ex:abc   ex:def\n" +
+				"                           ]\n" +
```

Comment:
I am not sure this one really counts as a bug, to be honest - and even if it does it is a _very_ exotic corner case. 

This case is an invalid collection structure in the sense that it links a non-list related poperty (`ex:abc`) to a bnode that that is directly part of the list structure itself. We are not so much inlining that blank node as just removing it, because the list gets converted to a syntactical shortform where the blank nodes and (rdf:first/rest properties) are just erased.

The actual output for this is:

```
@prefix ex: <http://example.com/ns#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .

ex:ex1 ex:list ([]) .
[] ex:abc ex:def .
```

In other words, one list with a single item (a blank node), and a statement with a blank node subject. The fact that in the original file that blank node was the _same_ blank node as the one used in the internal collection structure is lost, but that does not make the output incorrect. There's also a strong whiff of "what the heck are you even _doing_ using a collection blank node like this?!"

We could potentially handle this with a separate test to check if the collection is well-formed before serializing it, but it's going to be costly.

Reply:
I know it's pretty funky. I have no idea how it got produced to be honest. I still think we should consider it a bug though.

Reply:
I think the output is incorrect. The pretty printer and bnode inliner should return the same data if you first output n-triples, then pretty print and inline it, then convert it to n-triples again.

Reply:
> I think the output is incorrect. The pretty printer and bnode inliner should return the same data if you first output n-triples, then pretty print and inline it, then convert it to n-triples again

My point is that effectively, it does.